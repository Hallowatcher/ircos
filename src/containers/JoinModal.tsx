
import * as React from 'react';
import Modal from 'react-modal';

interface IState {
  isOpen: boolean;
  onClose: () => void;
  joinChannel: (event: any) => void;
}

const styles = {
  content: {
    top: '50%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    marginRight: '-50%',
    transform: 'translate(-50%, -50%)'
  }
};

export class JoinModal extends React.Component<IState, any> {
  constructor() {
    super();

    this.state = {
      channel: ''
    };

    this.joinChannel = this.joinChannel.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  public render() {
    return (
      <Modal
        isOpen={this.props.isOpen}
        onRequestClose={this.props.onClose}
        contentLabel='Join channel'
        style={styles}
      >
        <h1>Join channel modal</h1>
        <form onSubmit={this.joinChannel}>
          <input placeholder="#channel or user" onChange={this.handleChange} />
          <input type="submit" />
        </form>
      </Modal>
    );
  }

  private handleChange(event: any) {
    this.setState({ channel: event.target.value });
  }

  private joinChannel(event: any) {
    event.preventDefault();

    if (this.state.channel !== '') {
      this.props.joinChannel(this.state.channel);
      this.props.onClose();
    }
  }
}
