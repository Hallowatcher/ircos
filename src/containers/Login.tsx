import * as React from 'react';
import { connect } from 'react-redux';
import { LoginForm } from '../components/LoginForm';
import { createConnection } from '../actions/client';

type Props = {
  createConnection: any
}

const divStyle: React.CSSProperties = {
  margin: 0,
  padding: 0,
  width: '100%',
  height: '100%'
}

export class Login extends React.Component<Props, null> {
  render() {
    return (
      <div style={divStyle}>
        <LoginForm submitLogin={this.props.createConnection.bind(this)} />
      </div>
    )
  }
}

/* istanbul ignore next */
function dispatchToProps(dispatch: any) {
   return {
    createConnection: (user: string, pass: string) => { dispatch( createConnection(user, pass) ) }
  }
}

/* istanbul ignore next */
export default connect<any, any, any>(null, dispatchToProps)(Login);