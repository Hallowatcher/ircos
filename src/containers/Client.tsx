import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { connect } from 'react-redux';

import {
  makeCurrentChannel,
  sendMessage,
  leaveChannel,
  join
} from '../actions/client';

import { Tab } from '../components/Tab';
import { AddTab } from '../components/AddTab';
import { ChatView } from '../components/ChatView';
import { JoinModal } from './JoinModal';

type StateProps = {
  channels: any,
  messages: any[],
  currentChannel: string,
  nick: string,
  userId: number,
  channelLength: number
}

type DispatchProps = {
  makeCurrentChannel: (channel: string) => void,
  sendMessage: (channel: string, message: string) => void,
  closeChannel: (channel: string) => void ,
  joinChannel: (channel: string) => void
}

let styles = {
  container: {
    display: 'flex',
    flexDirection: 'column',
    width: '100%',
    height: '100%',
    overflow: 'hidden'
  } as React.CSSProperties,
  topBar: {
    backgroundColor: '#FE4590',
    flexShrink: 0,
    display: 'flex',
    flexDirection: 'row'
  } as React.CSSProperties,
  tabs: {
    flexGrow: 1,
    padding: '0 5px',
    display: 'flex',
    overflowY: 'hidden',
    overflowX: 'auto',
    alignItems: 'flex-end'
  } as React.CSSProperties,
  currentUser: {
    backgroundColor: 'rgba(255, 255, 255, 0.4)',
    display: 'flex',
    color: 'white'
  } as React.CSSProperties,
  currentUserImageContainer: {
    height: '35px',
    width: '35px',
    backgroundColor: 'rgba(255, 255, 255, 0.8)'
  } as React.CSSProperties,
  currentUserImage: {
    maxHeight: '100%'
  } as React.CSSProperties,
  currentUserNick: {
    padding: 5,
    display: 'flex',
    alignItems: 'center'
  } as React.CSSProperties,
  chat: {
    flexGrow: 1,
    overflowY: 'auto',
    overflowX: 'hidden',
    wordBreak: 'break-word'
  } as React.CSSProperties,
  dummyDiv: {
    float: 'left',
    clear: 'both'
  } as React.CSSProperties
}

export class Client extends React.Component<StateProps & DispatchProps, any> {

  constructor(props: any) {
    super(props);

    this.state = {
      msg: '',
      showJoinModal: false
    }
  }

  displayMessages = [];
  displayMessagesAmount: number = 50

  handleChangeMsg(event: any) {
    this.setState({ msg: event.target.value });
  }

  handleSubmit(event: any) {
    event.preventDefault();

    this.props.sendMessage(this.props.currentChannel, this.state.msg);
    this.setState({ msg: '' });
  }

  render() {

    // Tabs
    let channels = Object.keys(this.props.channels);
    let channelMap = channels.map((channel, index) => 
      <Tab
        key={index}
        tabName={channel}
        tabClick={this.props.makeCurrentChannel.bind(this)}
        closeTab={this.props.closeChannel.bind(this)}
        isActive={this.props.currentChannel === channel}
      />
    );

    let userImage = (() => {
      if (this.props.userId !== 0)
        return <img style={styles.currentUserImage} src={`https://a.ppy.sh/${this.props.userId}_${Date.now()}.jpg`} />;
    })();

    return (
      <div style={styles.container}>

        {/*Modals*/}
        <JoinModal
          isOpen={this.state.showJoinModal}
          onClose={() => { this.setState({ showJoinModal: false }) }}
          joinChannel={ this.props.joinChannel.bind(this) }
        />

        {/*Top bar*/}
        <div style={styles.topBar}>

          {/*Tabs*/}
          <div style={styles.tabs}>
            {channelMap}
            <AddTab clickAddTab={() => this.setState({ showJoinModal: true })} />
          </div>

          {/*Current user*/}
          <div style={styles.currentUser}>
            <div style={styles.currentUserImageContainer}>{userImage}</div>
            <div style={styles.currentUserNick}>{this.props.nick}</div>
          </div>

        </div>

        {/*Chat container*/}
        <ChatView
          style={styles.chat}
          messages={this.props.messages}
          selfNick={this.props.nick}
        />

        {/*Send message*/}
        <form onSubmit={this.handleSubmit.bind(this)}>
          <input type="text" placeholder="Enter a message..." value={this.state.msg} onChange={this.handleChangeMsg.bind(this)} />
          <input type="submit" value="Submit" />
        </form>
        
      </div>
    )
  }
}

/* istanbul ignore next */
function stateToProps(state: any) {
  return {
    channels: state.get('channelDb').toJS(),
    messages: state.getIn(['channelCurrent', 'messages']).toJS(),
    currentChannel: state.getIn(['channelCurrent', 'name']),
    nick: state.getIn(['userInfo', 'userName']),
    userId: state.getIn(['userInfo', 'userId']),
    channelLength: state.getIn(['settings', 'channelLength'])    
  }
}

/* istanbul ignore next */
function dispatchToProps(dispatch: any) {
  return {
    makeCurrentChannel: (channel: string) => { dispatch( makeCurrentChannel(channel) ) },
    sendMessage: (channel: string, message: string) => { dispatch( sendMessage(channel, message) ) },
    closeChannel: (channel: string) => { dispatch( leaveChannel(channel) ) },
    joinChannel: (channel: string) => { dispatch( join(channel) ) }
  }
}

/* istanbul ignore next */
export default connect(stateToProps, dispatchToProps)(Client);