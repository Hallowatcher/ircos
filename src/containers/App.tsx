import * as React from 'react';
import { connect } from 'react-redux';

import Login from './Login';
import Client from './Client';

type StateProps = {
  connected: boolean
}

export class App extends React.Component<StateProps, any> {

  render() {
    if (!this.props.connected) {
      return <Login />
    }

    return <Client />
  }
}

/* istanbul ignore next */
function stateToProps(state: any): StateProps {
  return {
    connected: !!state.getIn(['serverInfo', 'connected'])
  }
}

/* istanbul ignore next */
export default connect<StateProps, any, any>(stateToProps)(App);