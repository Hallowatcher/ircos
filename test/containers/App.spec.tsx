import * as React from 'react'
import { App } from '../../src/containers/App'
import * as enzyme from 'enzyme'

describe('Client', function () {

  let component: enzyme.ShallowWrapper<any, any>
  let connected: boolean
  beforeEach(function () {

    component = enzyme.shallow(
      <App connected={connected} />
    )
  })

  it('should render with login screen', function () {
    expect(component).toMatchSnapshot()
  })

  it('should render with client screen', function () {
    component.setProps({ connected: true })
    expect(component).toMatchSnapshot()
  })

})