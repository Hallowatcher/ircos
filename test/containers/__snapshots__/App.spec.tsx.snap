// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client should render with client screen 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Connect(Client) />,
  "nodes": Array [
    <Connect(Client) />,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <App
        connected={true}
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": App {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "connected": true,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Connect(Client) />,
        "_debugID": 5,
        "_renderedOutput": <Connect(Client) />,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <App
    connected={true}
/>,
}
`;

exports[`Client should render with login screen 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Connect(Login) />,
  "nodes": Array [
    <Connect(Login) />,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <App
        connected={undefined}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": App {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "connected": undefined,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Connect(Login) />,
        "_debugID": 2,
        "_renderedOutput": <Connect(Login) />,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <App
    connected={undefined}
/>,
}
`;
